File: src/core/crop_aspect.py
"""
Pure (Qt-free) geometry + presets backing the Crop panel's aspect-ratio and
straighten features. Kept import-light and side-effect-free so it can be unit
tested without a GUI. See spec/crop-panel.md.

Ratios are width/height in *Landscape* orientation; `oriented_ratio` applies the
portrait/landscape toggle, and `effective_box_ratio` maps an on-screen target
ratio onto the un-rotated image space the crop box lives in (90/270 coarse
rotation swaps width<->height).
"""

# (label, key, base_ratio): base_ratio is width/height for Landscape, or None
# for Free, or the sentinel "original" (resolve from the image's own aspect).
ASPECT_PRESETS = [
    ("Free", "free", None),
    ("Original", "original", "original"),
    ("1:1", "1:1", 1.0),
    ("5:4", "5:4", 5.0 / 4.0),
    ("4:3", "4:3", 4.0 / 3.0),
    ("7:5", "7:5", 7.0 / 5.0),
    ("3:2", "3:2", 3.0 / 2.0),
    ("16:9", "16:9", 16.0 / 9.0),
    ("Custom…", "custom", None),
]

# Keys for which the Landscape/Portrait toggle is meaningless.
ORIENTATION_FIXED_KEYS = {"free", "original", "1:1", "custom"}


def preset_for_key(key):
    """Return the (label, key, base_ratio) tuple for `key`, or None."""
    for entry in ASPECT_PRESETS:
        if entry[1] == key:
            return entry
    return None


def oriented_ratio(ratio, landscape):
    """Apply the orientation toggle to a base (Landscape) width/height ratio.
    Portrait flips it to height/width. None / non-positive pass through."""
    if ratio is None or ratio <= 0:
        return ratio
    return ratio if landscape else 1.0 / ratio


def effective_box_ratio(display_ratio, rotation):
    """Convert an on-screen target ratio into the box-frame ratio used on the
    pending crop rect (which lives in un-rotated image space). A 90/270 coarse
    rotation swaps width<->height as seen on screen; flips don't change a
    rectangle's aspect."""
    if display_ratio is None or display_ratio <= 0:
        return display_ratio
    if rotation % 180 == 90:
        return 1.0 / display_ratio
    return display_ratio


def enforce_ratio_size(adx, ady, r):
    """Smallest box of ratio r = w/h that COVERS the desired extents
    (|adx|, |ady|). Used while drawing/corner-dragging so the box grows to
    include the pointer without breaking the ratio. r None/<=0 -> unconstrained."""
    adx, ady = abs(adx), abs(ady)
    if r is None or r <= 0:
        return adx, ady
    if adx >= ady * r:
        return adx, adx / r
    return ady * r, ady


def fit_ratio_within(w, h, r):
    """Largest box of ratio r = w/h that FITS inside w x h (centered use).
    r None/<=0 or degenerate bounds -> (w, h) unchanged."""
    if r is None or r <= 0 or w <= 0 or h <= 0:
        return w, h
    if w / h >= r:        # height is the binding dimension
        return h * r, h
    return w, w / r       # width is the binding dimension


def reshape_to_ratio(center, box_wh, img_wh, r):
    """Compute a box of ratio r for a preset change / auto-seed.

    - center: (cx, cy) of the current box, in image px (ignored when box_wh is
      falsy — the new box is centered on the image instead).
    - box_wh: (w, h) of the current pending box, or None to seed from the image.
    - img_wh: (W, H) of the full image.
    - r: target box-frame ratio (w/h), or None for Free.

    Returns (cx, cy, bw, bh), or None for Free / degenerate input. When a box
    exists the result is fit *within* it (never larger than the current
    selection); otherwise it is the largest centered box of ratio r in the image.
    """
    if r is None or r <= 0:
        return None
    W, H = img_wh
    if W <= 0 or H <= 0:
        return None
    if box_wh is not None and box_wh[0] >= 2 and box_wh[1] >= 2:
        cx, cy = center
        bw, bh = fit_ratio_within(box_wh[0], box_wh[1], r)
    else:
        cx, cy = W / 2.0, H / 2.0
        bw, bh = fit_ratio_within(W, H, r)
    return cx, cy, bw, bh


def folded_crop_angle(crop_angle, fine_rotation_angle):
    """The crop-box angle that absorbs an image-level micro-rotation, so the two
    rotations don't stack. `fine_rotation_angle` is in hundredths of a degree
    (CCRImage convention). Equivalence (see spec §5.4): image fine-rotation
    rotates content clockwise for positive values, the crop box de-rotates
    counter-clockwise, hence the sign flip."""
    return (crop_angle or 0.0) - (fine_rotation_angle or 0) / 100.0

File: native/Sources/FreeCCRNative/CropAspect.swift
import CoreGraphics

/// The Crop panel's aspect-ratio presets — ported from `crop_aspect.py`'s
/// `ASPECT_PRESETS`. `Custom…` (manual W:H spinboxes) isn't in this port yet
/// — presets only, see `CropAspect`'s doc comment for the rest of what's
/// deferred.
enum CropAspectKey: String, CaseIterable, Identifiable {
    case free, original
    case oneToOne, fiveToFour, fourToThree, sevenToFive, threeToTwo, sixteenToNine
    case academy, flat, univisium, cinemaScope, dciScope

    var id: String { rawValue }

    var label: String {
        switch self {
        case .free: return "Free"
        case .original: return "Original"
        case .oneToOne: return "1:1"
        case .fiveToFour: return "5:4"
        case .fourToThree: return "4:3"
        case .sevenToFive: return "7:5"
        case .threeToTwo: return "3:2"
        case .sixteenToNine: return "16:9"
        case .academy: return "Academy (1.37:1)"
        case .flat: return "1.85:1 (Flat)"
        case .univisium: return "2:1 (Univisium)"
        case .cinemaScope: return "2.35:1 (CinemaScope)"
        case .dciScope: return "2.39:1 (Scope)"
        }
    }

    /// Width/height ratio in LANDSCAPE orientation; `nil` for `.free` and for
    /// `.original` (resolved dynamically from the loaded image's own aspect
    /// — see `CropAspect.effectiveRatio`), matching `crop_aspect.py`'s
    /// `base_ratio` column (`None` / the `"original"` sentinel).
    var baseRatio: Double? {
        switch self {
        case .free, .original: return nil
        case .oneToOne: return 1.0
        case .fiveToFour: return 5.0 / 4.0
        case .fourToThree: return 4.0 / 3.0
        case .sevenToFive: return 7.0 / 5.0
        case .threeToTwo: return 3.0 / 2.0
        case .sixteenToNine: return 16.0 / 9.0
        case .academy: return 1.375
        case .flat: return 1.85
        case .univisium: return 2.0
        case .cinemaScope: return 2.35
        case .dciScope: return 2.39
        }
    }

    /// Keys for which the Landscape/Portrait toggle is meaningless — mirrors
    /// `crop_aspect.py`'s `ORIENTATION_FIXED_KEYS` (minus `"custom"`, not
    /// ported).
    var isOrientationFixed: Bool {
        switch self {
        case .free, .original: return true
        default: return false
        }
    }
}

/// Pure (no Python, no AppKit) geometry backing the Crop panel — ported from
/// `crop_aspect.py`. This port only covers PRESET selection (pick a ratio,
/// get a box centered on the image) — not `crop_aspect.py`'s full feature
/// set: no draggable corner/edge handles (`enforce_ratio_size`), no "fit
/// within the CURRENT box" behavior (`reshape_to_ratio`'s box_wh branch —
/// this always reseeds centered on the full image, since there's no
/// existing draggable box to fit within yet), and no fine-rotation folding
/// (`folded_crop_angle` — this port has no separate "fine rotation" concept,
/// so the Straighten slider writes `crop_angle` directly). All boxes are
/// centered; a future milestone can add drag-to-move/resize on top of this.
enum CropAspect {
    /// The actual ratio to crop to, resolving `.original` against the real
    /// image size and applying the landscape/portrait toggle (`.free` stays
    /// `nil` throughout — no crop).
    static func effectiveRatio(for key: CropAspectKey, landscape: Bool, imageSize: CGSize) -> Double? {
        switch key {
        case .free:
            return nil
        case .original:
            guard imageSize.width > 0, imageSize.height > 0 else { return nil }
            return Double(imageSize.width / imageSize.height)
        default:
            return orientedRatio(key.baseRatio, landscape: landscape)
        }
    }

    /// Applies the orientation toggle to a base (Landscape) ratio — Portrait
    /// flips it to height/width. `nil`/non-positive pass through unchanged.
    static func orientedRatio(_ ratio: Double?, landscape: Bool) -> Double? {
        guard let ratio, ratio > 0 else { return ratio }
        return landscape ? ratio : 1.0 / ratio
    }

    /// The largest box of ratio `r = w/h` that fits inside `w x h`. `r`
    /// `nil`/non-positive or degenerate bounds return `(w, h)` unchanged.
    static func fitRatioWithin(_ w: Double, _ h: Double, _ r: Double?) -> (width: Double, height: Double) {
        guard let r, r > 0, w > 0, h > 0 else { return (w, h) }
        if w / h >= r { return (h * r, h) } // height is the binding dimension
        return (w, w / r) // width is the binding dimension
    }

    /// A normalized (0...1 fractions of the image, matching
    /// `CCRImage.crop_rect`'s coordinate convention) box for `key`, centered
    /// on the image — `nil` for `.free` (no crop) or before an image is
    /// loaded.
    static func normalizedRect(for key: CropAspectKey, landscape: Bool, imageSize: CGSize) -> CGRect? {
        guard let ratio = effectiveRatio(for: key, landscape: landscape, imageSize: imageSize),
              imageSize.width > 0, imageSize.height > 0 else { return nil }
        let (boxWidth, boxHeight) = fitRatioWithin(Double(imageSize.width), Double(imageSize.height), ratio)
        let w = boxWidth / Double(imageSize.width)
        let h = boxHeight / Double(imageSize.height)
        return CGRect(x: (1 - w) / 2, y: (1 - h) / 2, width: w, height: h)
    }
}
